sql/colexec: TestRouterOutputNext failed

[(sql/colexec).TestRouterOutputNext failed](https://viewLog.html?buildId=1697835&tab=buildLog) on [master@13af75f3d1376fff31d1419df2951c2d8a810efd](https://github.com/cockroachdb/cockroach/commits/13af75f3d1376fff31d1419df2951c2d8a810efd):

```
=== RUN   TestRouterOutputNext
--- FAIL: TestRouterOutputNext (0.01s)
=== RUN   TestRouterOutputNext/AddBatchDoesntBlockWhenOutputIsBlocked
    --- FAIL: TestRouterOutputNext/AddBatchDoesntBlockWhenOutputIsBlocked (0.00s)
        routers_test.go:283: unexpectedly still unblocked

```

<details><summary>Repro</summary><p>
Parameters:

- GOFLAGS=-json

```
make stressrace TESTS=TestRouterOutputNext PKG=./pkg/sql/colexec TESTTIMEOUT=5m STRESSFLAGS='-timeout 5m' 2>&1
```

[roachdash](https://roachdash.crdb.dev/?filter=status%3Aopen+t%3A.%2ATestRouterOutputNext.%2A&sort=title&restgroup=false&display=lastcommented+project)

<sub>powered by [pkg/cmd/internal/issues](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)</sub></p></details>

[(sql/colexec).TestRouterOutputNext failed](https://viewLog.html?buildId=1710174&tab=buildLog) on [master@e88ba158866f940dc4c59a23bcd1a4467db439be](https://github.com/cockroachdb/cockroach/commits/e88ba158866f940dc4c59a23bcd1a4467db439be):

```
=== RUN   TestRouterOutputNext
--- FAIL: TestRouterOutputNext (0.01s)
=== RUN   TestRouterOutputNext/AddBatchDoesntBlockWhenOutputIsBlocked
    --- FAIL: TestRouterOutputNext/AddBatchDoesntBlockWhenOutputIsBlocked (0.00s)
        routers_test.go:283: unexpectedly still unblocked

```

<details><summary>More</summary><p>
Parameters:

- GOFLAGS=-json

```
make stressrace TESTS=TestRouterOutputNext PKG=./pkg/sql/colexec TESTTIMEOUT=5m STRESSFLAGS='-timeout 5m' 2>&1
```


[See this test on roachdash](https://roachdash.crdb.dev/?filter=status%3Aopen+t%3A.%2ATestRouterOutputNext.%2A&sort=title&restgroup=false&display=lastcommented+project)
<sub>powered by [pkg/cmd/internal/issues](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)</sub></p></details>

--- a/pkg/sql/colexec/routers_test.go
+++ b/pkg/sql/colexec/routers_test.go
@@ -238,19 +238,20 @@ func TestRouterOutputNext(t *testing.T) {
 	})
 
 	t.Run("AddBatchDoesntBlockWhenOutputIsBlocked", func(t *testing.T) {
-		const (
+		var (
 			smallBatchSize = 8
 			blockThreshold = smallBatchSize / 2
 		)
 
-		// It is possible that coldata.BatchSize is smaller than our smallBatchSize
-		// which breaks the assumptions of this test. Also, interestingly, if we
-		// do something like 'if smallBatchSize < coldata.BatchSize()', apparently
-		// the compiler will produce such code that the 'if' condition will be only
-		// checked on the first run whereas the condition might not longer hold on
-		// consequent runs. That's why we have this "data dependency" on the slice.
-		if len(fullSelection) < blockThreshold {
-			return
+		if len(fullSelection) <= smallBatchSize {
+			// If a full batch is smaller than our small batch size, reduce it, since
+			// this test relies on multiple batches returned from the input.
+			smallBatchSize = 2
+			if smallBatchSize >= coldata.MinBatchSize {
+				// Sanity check.
+				t.Fatalf("smallBatchSize=%d still too large (must be less than MinBatchSize=%d)", smallBatchSize, coldata.MinBatchSize)
+			}
+			blockThreshold = 1
 		}
 
 		// Use a smaller selection than the batch size; it increases test coverage.
@@ -267,7 +268,7 @@ func TestRouterOutputNext(t *testing.T) {
 		o := newRouterOutputOpWithBlockedThresholdAndBatchSize(
 			testAllocator, []coltypes.T{coltypes.Int64}, ch, blockThreshold, int(coldata.BatchSize()),
 		)
-		in := newOpTestInput(smallBatchSize, data, nil /* typs */)
+		in := newOpTestInput(uint16(smallBatchSize), data, nil /* typs */)
 		out := newOpTestOutput(o, expected)
 		in.Init()
 

